File: frontend/src/api/overview.ts
import { getJson } from './http-client';

export interface OverviewTimelinePoint {
  readonly period_start: string;
  readonly actual_units: number | null;
  readonly forecast_units: number | null;
}

export interface OverviewRiskCounts {
  readonly stockout: number;
  readonly below_safety_stock: number;
  readonly healthy: number;
  readonly excess: number;
  readonly unavailable: number;
}

export interface OverviewException {
  readonly exception_id: string;
  readonly store_id: string;
  readonly store_name: string;
  readonly product_id: string;
  readonly sku: string;
  readonly product_name: string;
  readonly category: string;
  readonly exception_type:
    | 'potential_stockout'
    | 'below_safety_stock'
    | 'excess_inventory'
    | 'forecast_uncertainty'
    | 'forecast_bias';
  readonly severity: 'medium' | 'high' | 'critical';
  readonly priority_score: number;
  readonly title: string;
  readonly description: string;
  readonly metric_name: string;
  readonly metric_value: number;
  readonly threshold_value: number;
  readonly selected_model: string;
}

export interface PortfolioOverview {
  readonly run_id: string;
  readonly data_version: string;
  readonly training_cutoff: string;
  readonly horizon_months: number;
  readonly store_id: string | null;
  readonly store_name: string;
  readonly series_count: number;
  readonly recent_12_month_demand: number;
  readonly prior_12_month_demand: number;
  readonly demand_change: number | null;
  readonly forecast_horizon_demand: number;
  readonly median_series_wape: number | null;
  readonly median_series_mase: number | null;
  readonly median_absolute_bias: number | null;
  readonly weighted_interval_coverage: number | null;
  readonly risk_counts: OverviewRiskCounts;
  readonly timeline: readonly OverviewTimelinePoint[];
  readonly priority_exceptions: readonly OverviewException[];
}

export function getPortfolioOverview(
  runId: string,
  storeId: string,
  signal?: AbortSignal,
): Promise<PortfolioOverview> {
  const parameters = new URLSearchParams({ exception_limit: '10' });
  if (runId) parameters.set('run_id', runId);
  if (storeId) parameters.set('store_id', storeId);
  return getJson<PortfolioOverview>(`/api/v1/dashboard/summary?${parameters.toString()}`, signal);
}

File: backend/app/domain/overview.py
"""Storage-independent portfolio overview read models."""

from dataclasses import dataclass
from datetime import date
from typing import Literal

OverviewExceptionType = Literal[
    "potential_stockout",
    "below_safety_stock",
    "excess_inventory",
    "forecast_uncertainty",
    "forecast_bias",
]
OverviewExceptionSeverity = Literal["medium", "high", "critical"]


@dataclass(frozen=True, slots=True)
class OverviewDemandPoint:
    period_start: date
    actual_units: int | None
    forecast_units: float | None


@dataclass(frozen=True, slots=True)
class OverviewSeriesEvidence:
    store_id: str
    store_name: str
    product_id: str
    sku: str
    product_name: str
    category: str
    selected_model: str
    wape: float | None
    mase: float | None
    bias: float | None
    interval_coverage: float | None
    validation_points: int | None
    available_inventory: int
    on_order_due_within_lead_time: int
    safety_stock: int
    lead_time_days: int
    covered_lead_time_days: int
    forecast_demand_during_lead_time: float
    average_monthly_forecast: float
    average_interval_width: float


@dataclass(frozen=True, slots=True)
class OverviewException:
    exception_id: str
    store_id: str
    store_name: str
    product_id: str
    sku: str
    product_name: str
    category: str
    exception_type: OverviewExceptionType
    severity: OverviewExceptionSeverity
    priority_score: int
    title: str
    description: str
    metric_name: str
    metric_value: float
    threshold_value: float
    selected_model: str


@dataclass(frozen=True, slots=True)
class OverviewRiskCounts:
    stockout: int
    below_safety_stock: int
    healthy: int
    excess: int
    unavailable: int


@dataclass(frozen=True, slots=True)
class PortfolioOverview:
    run_id: str
    data_version: str
    training_cutoff: date
    horizon_months: int
    store_id: str | None
    store_name: str
    series_count: int
    recent_12_month_demand: int
    prior_12_month_demand: int
    demand_change: float | None
    forecast_horizon_demand: float
    median_series_wape: float | None
    median_series_mase: float | None
    median_absolute_bias: float | None
    weighted_interval_coverage: float | None
    risk_counts: OverviewRiskCounts
    timeline: tuple[OverviewDemandPoint, ...]
    priority_exceptions: tuple[OverviewException, ...]
